lxd/device/gpu_physical.go

```diff
@@ -98,10 +101,209 @@ func (d *gpuPhysical) Start() (*deviceConfig.RunConfig, error) {
+		fsName := "none"
+
+		fileInfo, err := os.Stat(srcPath)
+		if err != nil {
+			return fmt.Errorf("Failed accessing source path %q: %w", srcPath, err)
+		}
+
+		fileMode := fileInfo.Mode()
+		isFile := false
+		if !fileMode.IsDir() {
+			isFile = true
+		}
+
+		f, err := os.OpenFile(srcPath, unix.O_PATH|unix.O_CLOEXEC, 0)
+		if err != nil {
+			return fmt.Errorf("Failed opening source path %q: %w", srcPath, err)
+		}
+
+		srcPath = fmt.Sprintf("/proc/self/fd/%d", f.Fd())
+		srcFDHandlers = append(srcFDHandlers, f)
+
+		// Create the devices directory if missing.
+		if !shared.PathExists(d.inst.DevicesPath()) {
+			err := os.Mkdir(d.inst.DevicesPath(), 0711)
+			if err != nil {
+				return err
+			}
+		}
+
+		// Clean any existing entry.
+		// We could have just continued because this qualifies as a duplicate case
+		// (e.g: trying to mount the same user space mount) but then if we did that
+		// we would stop the container from receiving a potentially 'updated' version
+		// of that mount (e.g: auto-update a user space lib in the container)
+		// We prefer that approach in order to keep the host and container user spaces in sync.
+		if shared.PathExists(devPath) {
+			err := os.Remove(devPath)
+			if err != nil {
+				return err
+			}
+		}
+
+		// Create the mount point.
+		if isFile {
+			f, err := os.Create(devPath)
+			if err != nil {
+				return err
+			}
+
+			srcFDHandlers = append(srcFDHandlers, f)
+		} else {
+			err := os.Mkdir(devPath, 0700)
+			if err != nil {
+				return err
+			}
+		}
+
+		// Mount the fs.
+		err = DiskMount(srcPath, devPath, false, "", mntOptions, fsName)
+		if err != nil {
+			return err
+		}
+
+		if isFile {
+			options = append(options, "create=file")
+		} else {
+			options = append(options, "create=dir")
+		}
+
+		runConf.Mounts = append(runConf.Mounts, deviceConfig.MountEntryItem{
+			DevName:    deviceName,
+			DevPath:    devPath,
+			TargetPath: relativeDestPath,
+			FSType:     "none",
+			Opts:       options,
+			OwnerShift: ownerShift,
+		})
+
+		runConf.PostHooks = append(runConf.PostHooks, func() error {
+			err := unix.Unmount(devPath, unix.MNT_DETACH)
+			if err != nil {
+				return err
+			}
+
+			return nil
+		})
+	}
+
+	return nil
+}
+
+func (d *gpuPhysical) generateCDIHooksFilePath() string {
+	return filepath.Join(d.inst.DevicesPath(), fmt.Sprintf("%s%s", d.name, cdi.CDIHooksFileSuffix))
+}
+
 // startContainer detects the requested GPU devices and sets up unix-char devices.
 // Returns RunConfig populated with mount info required to pass the unix-char devices into the container.
 func (d *gpuPhysical) startContainer() (*deviceConfig.RunConfig, error) {
 	runConf := deviceConfig.RunConfig{}
+	if d.config["id"] != "" {
+		// Check if the id of the device match a CDI format.
+		cdiID, err := cdi.ToCDI(d.config["id"])
+		if err != nil {
+			return nil, err
+		}
+
+		// The cdiID can be empty if the id doesn't match the general CDI format, but this is not an error.
```

Comment:
Comment should be improved here to indicate we are using the CDI approach.

Reply:
fixed